File: framework/scala/src/main/scala/soda/scala/unittest/Validatable.scala
package soda.scala.unittest

import play.api.libs.json.Json

import scala.reflect.runtime.universe.Type

trait Validatable[R] {

  var compareSerial: Boolean = false

  protected var validator: Option[(R, R) => Boolean] = None

  var expectedOutput: Option[R] = None

  def validate(input: WorkInput, retType: Type, result: R, elapseMillis: Double): WorkOutput = {
    val resConv = ConverterFactory.create(retType).asInstanceOf[ObjectConverter[R]]
    val serialResult = resConv.toJsonSerializable(result)
    val output = new WorkOutput
    output.id = input.id
    output.result = serialResult
    output.elapse = elapseMillis

    var success = true
    if (input.hasExpected) {
      if (compareSerial && validator.isEmpty) {
        val a = Json.stringify(input.expected)
        val b = Json.stringify(serialResult)
        success = a == b
      } else {
        val expect = resConv.fromJsonSerializable(input.expected)
        expectedOutput = Some(expect)
        success = validator match {
          case Some(va) => va.apply(expect, result)
          case None => ValidatorFactory.create(retType)(expect, result)
        }
      }
    }
    output.success = success
    output
  }

}

File: framework/java/src/main/java/soda/unittest/Validatable.java
package soda.unittest;

import soda.unittest.conv.ConverterFactory;
import soda.unittest.conv.ObjectConverter;
import soda.unittest.validate.ValidatorFactory;

import java.lang.reflect.Type;
import java.util.Objects;
import java.util.function.BiPredicate;

class Validatable<R> {

    public boolean compareSerial;

    public BiPredicate<R,R> validator;

    public R expectedOutput;

    public void validate(WorkInput input, Type retType, R result, WorkOutput output) {
        Utils.wrapEx(() -> {
            ObjectConverter<R, Object> resConv = Utils.cast(ConverterFactory.createConverter(retType));
            output.result = resConv.toJsonSerializable(result);
            boolean success = true;
            if (input.expected != null) {
                if (compareSerial && validator == null) {
                    String a = Utils.objectMapper.writeValueAsString(input.expected);
                    String b = Utils.objectMapper.writeValueAsString(output.result);
                    success = Objects.equals(a, b);
                } else {
                    var expect = resConv.fromJsonSerializable(input.expected);
                    expectedOutput = expect;
                    if (validator != null) {
                        success = validator.test(expect, result);
                    } else {
                        success = ValidatorFactory.create(retType).test(expect, result);
                    }
                }
            }
            output.success = success;
        });
    }

}
